components/core/src/clp/BoundedReader.cpp

```diff
@@ -0,0 +1,39 @@
+#include "BoundedReader.hpp"
+
+namespace clp {
+auto BoundedReader::try_seek_from_begin(size_t pos) -> ErrorCode {
+    auto next_pos = pos > m_bound ? m_bound : pos;
+    if (auto const rc = m_reader->try_seek_from_begin(next_pos); ErrorCode_Success != rc) {
+        m_curr_pos = ErrorCode_EndOfFile == rc ? next_pos : m_curr_pos;
+        return rc;
+    }
+    m_curr_pos = next_pos;
+    if (m_curr_pos >= m_bound) {
+        return ErrorCode_EndOfFile;
+    }
+    return ErrorCode_Success;
```

Comment:
Just thinking out loud, in an ideal world, `rc = m_reader->try_seek_from_begin(next_pos)` should guarantee that `m_curr_pos >= m_bound`. So the code can be rewritten as the follow.

It doesn't really simplify the code a lot though, so it's totally up to you to choose which version to use.

```suggestion
    if (auto const rc = m_reader->try_seek_from_begin(next_pos); 
        ErrorCode_Success != rc && ErrorCode_EndOfFile != rc) {
        return rc;
    }
    m_curr_pos = next_pos;
    if (m_curr_pos >= m_bound) {
        return ErrorCode_EndOfFile;
    }
    return ErrorCode_Success;
```

Reply:
I think this one I'll leave because I have to think about this version of the code for an extra second before deciding that it works.